### ISSUE TITLE
Problems with InternalEntityTypeBuilder.IsIgnored(string, ConfigurationSource?) 

### ISSUE THREAD

**Issue by lajones** [dotnet/efcore#21220]
`IsIgnored` is implemented like this:

```C#
        public virtual bool IsIgnored([NotNull] string name, ConfigurationSource? configurationSource)
        {
            Check.NotEmpty(name, nameof(name));

            var ignoredConfigurationSource = Metadata.FindIgnoredConfigurationSource(name);
            return !configurationSource.HasValue
                || !configurationSource.Value.Overrides(ignoredConfigurationSource);
        }
```

I think it's meant to answer the question "is this property marked ignored at level `configurationSource` or above?". If so some problems are:
- if you pass `null` as `configurationSource` this method will return `true` regardless of whether a property with that name exists, let alone whether it is ignored.
- if you pass `Explicit` as `configurationSource` this method will return `false` regardless of whether a property with that name exists, let alone whether it is ignored.
- Suppose you have a Property which has the `[NotMapped]` attribute on it and another convention based on a `DataAnnotation` runs (e.g. `IndexAttributeAnnotation`). Then `ignoredConfigurationSource ` is `DataAnnotation`, but `configurationSource` is also `DataAnnotation` - `DataAnnotation` does "override" `DataAnnotation` - so this returns `false`, but the property did have the `[NotMapped]` attribute on it?

I think it should be:

```C#
        public virtual bool IsIgnored([NotNull] string name, ConfigurationSource? configurationSource)
        {
            Check.NotEmpty(name, nameof(name));

            var ignoredConfigurationSource = Metadata.FindIgnoredConfigurationSource(name);
            return ignoredConfigurationSource == null
                ? false
                : ignoredConfigurationSource.Overrides(configurationSource.Value);
        }
```

However doing that causes a number of tests to fail.

This issue is to check whether my statement above is correct and if so to do the work to fix the code. (Once that's done the workarounds for this in `IndexAttributeConvention` can be removed).

**Comment by smitpatel** [comment]
> is this property marked ignored at level configurationSource or above?

The actual question is -> I have `configurationSource` (which probably wants to add/use a "property"), is the "property" ignored by same configuration source or higher? (in which case I cannot get the property).
- I am bit unsure about 'null' case, it may be consistent with rest of the metadata in terms of how to treat null configuration source.
- Explicit case is correct. Nothing is ignored when the configuration source is Explicit since Explicit will override any other configuration which ignored the property.
- DataAnnotation case is also correct. Same level configuration source can override. `NotMapped` will get overriden by `Index`

**Comment by lajones** [comment]
- I think we should document our assumptions.

You say:
> the question is ... is the "property" ignored by same configuration source or higher (in which case I cannot get the property).
1. Your question sounds like a `CanSetIsIgnored()` method - is this supposed to be that? Did we name it badly?
2. If the property is ignored at a given configuration source, why can't a later convention using that same configuration source reset it?

**Comment by smitpatel** [comment]
re 2 - you earlier said 

> Suppose you have a Property which has the [NotMapped] attribute on it and another convention based on a DataAnnotation runs (e.g. IndexAttributeAnnotation). Then ignoredConfigurationSource is DataAnnotation, but configurationSource is also DataAnnotation - DataAnnotation does "override" DataAnnotation - so this returns false, but the property did have the [NotMapped] attribute on it?

If this method returns false then IndexAttribute will create the index hence reset the ignored part. Is it not working correctly then?

**Comment by lajones** [comment]
First I'd like to hone in on the question. Can I rephrase your question as "Can I set Ignore on this property at level `configurationSource`"? If so then a) we should rename it to `CanSetIgnore` or something similar, and b) I think we need to change it so that if you pass in `null` then the answer should be `true` if `ignoredConfigurationSource` is null, otherwise it should be `false`. For non-`null` `configurationSource` the answer should be `configurationSource.Value.Overrides(ignoredConfigurationSource)`.

If it had been called that I would never have been tempted to call that API as I want a different question answered.

**Comment by AndriySvyryd** [comment]
> Can I rephrase your question as "Can I set Ignore on this property at level configurationSource"? 

No, this method answers the question "Can I add an entity type member with the given name and configuration source?" We can rename it if that clears it up.

The bug in implementation is that it returns `false` for `DataAnnotation` if the member is ignored using `DataAnnotation`

**Comment by lajones** [comment]
TL;DR Yes I'd like to rename it. Sounds like a question for a Design Meeting.

But unfortunately again I find the phrasing of that question confusing. It depends upon the implication of what you mean by "add" if there is an existing member called `name`. If there is no such existing member then "Can I add an entity type member with the given name and configuration source?" makes sense (but I imagine would always be answered `true`). But if there is already a member called `name` then doesn't the question become "Can I override the setting of Ignored on that member using the given configuration source"? Or did you mean to imply something else?

I'm also still very confused about `configurationSource == null` cases. If there exists a property with name `name` and it has has been Ignored (i.e. `ignoredConfigurationSource != null`), I don't see how returning `true` to the question "Can I add an entity type member with the given name and configuration source `null`" is correct? Doesn't `null` rank lower than even `Convention` in the ranks of `ConfigurationSource`?

**Comment by AndriySvyryd** [comment]
Ok, to be more precise the question is "Is there an entity type member with the given name, if not can I add it with the given configuration source?"

Don't worry about the `configurationSource == null` case, it's not exposed on public API

**Comment by lajones** [comment]
> Don't worry about the `configurationSource == null` case, it's not exposed on public API

Since we don't document what these methods do, it makes it confusing if it doesn't work. The `null` case was the first case I looked at when I was trying to puzzle out what this method did. There's nothing to indicate that that is an invalid case. And it immediately didn't make any sense. Either we should document it and add a `Debug.Assert`, or if `configurationSource == null` really doesn't make any sense then we should change the method signature to make it non-optional.

**Comment by AndriySvyryd** [comment]
`configurationSource == null` is a valid case that is only used internally. It's not relevant to the external callers.

`IConventionEntityTypeBuilder.IsIgnored` is the documented public API.

**Comment by lajones** [comment]
I'm sorry, but I think we should rename the method and document what it is supposed to do. There have now been 3 different interpretations of what it is supposed to do - so clearly it is not obvious. I was asked to use this method (and not the public API). Its method name misled me into thinking I was asking a question about whether something was ignored, not about whether I could set something. Even after all the talks above I'm still not clear on what should happen in the `null` case and more importantly _why_ that should happen; what is the internal use case that's never exposed publicly, what does it mean? My being misled, my spending time trying to figure out why it was doing what it does, why it wasn't doing what I wanted/expected it to do and all the conversation above could have been avoided with a few comments.

In other words the documentation I'm asking for is not for the public, it's for maintainers/users of this code.

**Comment by AndriySvyryd** [comment]
> I'm sorry, but I think we should rename the method and document what it is supposed to do.

Feel free to bring this up in the design meeting if you think that spending more time on this is worth it.

>  I was asked to use this method (and not the public API).

This wasn't the case.

> Its method name misled me into thinking I was asking a question about whether something was ignored.

That __is__ what the method returns

>Even after all the talks above I'm still not clear on what should happen in the null case and more importantly why that should happen; what is the internal use case that's never exposed publicly, what does it mean?

Internally `configurationSource == null` means that the called method should return a positive value  only when the method will make no modifications to the model, for negative methods like `IsIgnored` this is inverted.

> In other words the documentation I'm asking for is not for the public, it's for maintainers/users of this code.

We should limit internal documentation to complex code that's not likely to change. The confusion here is that the method doesn't return the expected result due to a bug.
